consensus/tendermint/tendermint.go

```diff
@@ -0,0 +1,637 @@
+			proposalValue = &v
+		}
+		proposalMessage := Proposal[V, H, A]{
+			Height:     t.state.height,
+			Round:      r,
+			ValidRound: t.state.validRound,
+			Value:      proposalValue,
+			Sender:     t.nodeAddr,
+		}
+
+		t.messages.addProposal(proposalMessage)
+		fmt.Println("Broadcasting the proposal")
+		t.broadcasters.ProposalBroadcaster.Broadcast(proposalMessage)
+	} else {
+		t.scheduleTimeout(t.timeoutPropose(r), propose, t.state.height, t.state.round)
+	}
+
+	// Todo: write a new function startFutureRound which checks for line 22 and 28 incase prevote or precommit where
+	// the cause of the round change and the proposal for the future round was received prior future round change.
+	// Start a goroutine with which sends the messages to the
+}
+
+type timeout struct {
+	*time.Timer
+
+	s step
+	h uint
+	r uint
+}
+
+func (t *Tendermint[V, H, A]) scheduleTimeout(duration time.Duration, s step, h, r uint) {
+	tm := timeout{s: s, h: h, r: r}
+	tm.Timer = time.AfterFunc(duration, func() {
+		select {
+		case <-t.quit:
+		case t.timeoutsCh <- tm:
+		}
+	})
+	t.scheduledTms = append(t.scheduledTms, tm)
+}
+
+func (t *Tendermint[_, H, A]) OnTimeoutPropose(h, r uint) {
+	if t.state.height == h && t.state.round == r && t.state.step == propose {
+		vote := Prevote[H, A]{
+			Vote: Vote[H, A]{
+				Height: t.state.height,
+				Round:  t.state.round,
+				ID:     nil,
+				Sender: t.nodeAddr,
+			},
+		}
+		t.messages.addPrevote(vote)
+		fmt.Println("About to send prevot nil")
+		t.broadcasters.PrevoteBroadcaster.Broadcast(vote)
+		t.state.step = prevote
+		fmt.Println("Set the step to prevote")
+	}
+}
+
+func (t *Tendermint[_, H, A]) OnTimeoutPrevote(h, r uint) {
+	if t.state.height == h && t.state.round == r && t.state.step == prevote {
+		vote := Precommit[H, A]{
+			Vote: Vote[H, A]{
+				Height: t.state.height,
+				Round:  t.state.round,
+				ID:     nil,
+				Sender: t.nodeAddr,
+			},
+		}
+		t.messages.addPrecommit(vote)
+		fmt.Println("About to send precommit nil")
+		t.broadcasters.PrecommitBroadcaster.Broadcast(vote)
+		t.state.step = precommit
+		fmt.Println("Set the step to precommit")
+	}
+}
+
+func (t *Tendermint[_, _, _]) OnTimeoutPrecommit(h, r uint) {
+	if t.state.height == h && t.state.round == r {
+		t.startRound(r + 1)
+	}
+}
+
+func (t *Tendermint[V, H, A]) line55(futureR uint) {
+	vals := make(map[A]struct{})
+	proposals, prevotes, precommits := t.messages.allMessages(t.state.height, futureR)
+
+	// If a validator has sent proposl, prevote and precommit from a future round then it will only be counted once.
+	for addr := range proposals {
+		vals[addr] = struct{}{}
+	}
+
+	for addr := range prevotes {
+		vals[addr] = struct{}{}
+	}
+
+	for addr := range precommits {
+		vals[addr] = struct{}{}
+	}
+
+	if t.validatorSetVotingPower(slices.Collect(maps.Keys(vals))) > f(t.validators.TotalVotingPower(t.state.height)) {
+		t.startRound(futureR)
+	}
+}
+
+func (t *Tendermint[V, H, A]) validatorSetVotingPower(vals []A) uint {
```

Comment:
nit: rename something like `validatorsTotalVotingPower`, I keep reading this as a Set function. Maybe it's just me, so feel free to ignore :)

Reply:
Its personal preference, I will change the name.